File: Python/프로그래머스/3/42861. 섬 연결하기/섬 연결하기.py
import heapq

def solution(n, costs):
    ans = 0
    heap = []
    
    for v1, v2, c in costs:
        heapq.heappush(heap, (c, v1, v2))
        
    parents = [i for i in range(n)]
    
    def find(u, parents):
        
        if (u != parents[u]):
            parents[u] = find(parents[u], parents)
        
        return parents[u]
    
    def union(u, v, parents):
        pu = find(u, parents)
        pv = find(v, parents)
        
        if (pu == pv):
            return True
        
        if (pu > pv):
            parents[pv] = pu
        else:
            parents[pu] = pv
            
        return False
        
    while (heap):
        cost, v1, v2 = heapq.heappop(heap)
        # print(parents)
        if union(v1, v2, parents):
            continue
        
        ans += cost
    
    # print(parents)
    return ans

File: C++/프로그래머스/3/42861. 섬 연결하기/섬 연결하기.cpp
#include <bits/stdc++.h>

using namespace std;

int find(vector<int>& parents, int v){
    if (v != parents[v])
        parents[v] = find(parents, parents[v]);
    
    return parents[v];
}

bool union_(vector<int>& parents, int u, int v){
    int pu = find(parents, u);
    int pv = find(parents, v);
    
    if (pu == pv)
        return true;
    else if (pu < pv)
        parents[pv] = pu;
    else
        parents[pu] = pv;
    
    return false;
}


int solution(int n, vector<vector<int>> costs) {
    int ans = 0, num_edges = 0;
    vector<int> parents;
    
    for (int i=0; i < n; i++)
        parents.push_back(i);
    
    sort(costs.begin(), costs.end(), [](const vector<int> a, const vector<int> b){
        if (a[2] < b[2])
            return true;
        else if (a[2] > b[2])
            return false;
        else
            return a < b;
    });
    
    for (auto& cost: costs){
        int v1=cost[0], v2=cost[1], c=cost[2];
        
        if (union_(parents, v1, v2))
            continue;
        
        ans += c;
        num_edges += 1;
        if (num_edges == n-1)
            break;
    }
    
    return ans;
}